### DISCUSSION TITLE
Avoid writing converters

### DISCUSSION THREAD

**Discussion by nareshbhatia** [99designs/gqlgen#3086] [Q&A] [answered]
My team uses gqlgen for building our GraphQL server. The backend services expose their APIs using gRPC. What we are finding is that our GraphQL server is littered with converters for converting gRPC responses to gqlgen generated types – this, in spite of the fact that both objects map 1-to-1. Is there a better way?

I created a simple example to illustrate the problem.

The backend service defines a [Movie](https://github.com/nareshbhatia/movie-magic-services-go/blob/main/proto/models/movie/v1/movie.proto#L30-L44) object using protobuf:

```protobuf
message Movie {
  string id = 1;
  string name = 2;
  string description = 3;
  repeated CastMember cast = 4;
  Certificate certificate = 5;
  repeated string genres = 6;
  Image image = 7;
  int32 rank = 8;
  RatingsSummary ratings_summary = 9;
  int32 release_year = 10;
  int32 runtime = 11;
  string tagline = 12;
}
```

The GraphQL gateway defines the same [Movie](https://github.com/nareshbhatia/movie-magic-gateway-go/blob/main/graph/schema.graphql#L44-L57) object using a GraphQL schema:

```graphql
type Movie {
  id: ID!
  name: String!
  description: String!
  cast: [CastMember!]!
  certificate: Certificate!
  genres: [String!]!
  image: Image
  rank: Int!
  ratingsSummary: MovieRatingsSummary!
  releaseYear: Int!
  runtime: Int!
  tagline: String
}
```

While the objects are essentially the same, the movie resolver has to [convert the gRPC response to the GraphQL model](https://github.com/nareshbhatia/movie-magic-gateway-go/blob/main/pkg/graph/resolver/schema.resolvers.go#L25-L53):

```go
// Map the gRPC response to the GraphQL model
movies := make([]*model.Movie, len(res.Movies))
for i, movie := range res.Movies {
	movies[i] = &model.Movie{
		ID:   movie.Id,
		Name: movie.Name,
		Cast:        convertCastMembers(movie.Cast),
		Certificate: convertCertificate(movie.Certificate),
		Genres:      movie.Genres,
		Rank:        int(movie.Rank),
		RatingsSummary: &model.MovieRatingsSummary{
			AggregateRating: movie.RatingsSummary.AggregateRating,
			VoteCount:       int(movie.RatingsSummary.VoteCount),
		},
		ReleaseYear: int(movie.ReleaseYear),
	}
}
```

Not only that, there is [additional pain to convert enums](https://github.com/nareshbhatia/movie-magic-gateway-go/blob/main/pkg/graph/resolver/translate.go#L8-L25):

```go
func convertCertificate(certificate moviev1.Certificate) model.Certificate {
	switch certificate {
	case moviev1.Certificate_CERTIFICATE_UNSPECIFIED:
		return model.CertificateCertificateUnspecified
	case moviev1.Certificate_CERTIFICATE_G:
		return model.CertificateCertificateG
	case moviev1.Certificate_CERTIFICATE_NR:
		return model.CertificateCertificateNr
	case moviev1.Certificate_CERTIFICATE_PG_13:
		return model.CertificateCertificatePg13
	case moviev1.Certificate_CERTIFICATE_PG:
		return model.CertificateCertificatePg
	case moviev1.Certificate_CERTIFICATE_R:
		return model.CertificateCertificateR
	default:
		return model.CertificateCertificateUnspecified
	}
}
```

Is there a better workflow that I am not aware of?

**Comment by nareshbhatia** [comment]
Please help. Any insights on this?

**Comment by StevenACoffman** [comment, marked as answer]
Hi, so the same thing happens for any other codegen such as sqlc models. 

If the structs have identical fields, you can use Go's inbuilt struct conversion. 

In order to convert between mostly identical structs, you can use https://github.com/jmattheis/goverter , https://github.com/mitchellh/mapstructure or a few other libraries.

**Reply by nareshbhatia** [reply]
Thank you, @StevenACoffman. Goverter seems to be the closest match for this.
